File: frontend/src/api/tags.ts
import type { KnowledgeDocument } from './documents';
import type { components } from './generated';
import { request } from './client';

// ── 混合接入（openapi codegen · Slice A 试点）──────────────────────────
// 主体响应类型对齐生成类型（字段名 / 类型 / optionality 以 openapi 为准，消除手工双写）；
// openapi 几乎无 enum（仅 DocumentPermission / TermStatus），响应字段多为裸 string，
// 故 union literal 保留手写作 narrow overlay（保住 switch / 分支的编译期保护）。

/** Tag status —— openapi 是裸 `string`，保留手写 union 保编译期 narrow。 */
export type TagStatus = 'active' | 'archived';

/**
 * TagView —— 主体字段来自生成类型；`status` 字段 narrow 为手写 TagStatus
 *（Omit + 覆盖：后端加字段自动同步，status 保 union 保护）。
 */
export type TagView = Omit<components['schemas']['TagView'], 'status'> & {
  status: TagStatus;
};

/** DocumentTagView —— 与生成类型零差异（无 union），直接 alias。 */
export type DocumentTagView = components['schemas']['DocumentTagView'];

/** 文档↔标签关联视图（addDocumentTag 返回），与生成 TagLinkView 一致。 */
type TagLinkView = components['schemas']['TagLinkView'];
// ──────────────────────────────────────────────────────────────────────

/**
 * 分页容器。生成类型有 TagListPage / DocumentTagListPage，但其 `items` 元素是
 * 生成 TagView（status: string），与上面 narrow 后的 TagView 不匹配；保留手写泛型
 * 搭配 narrow 元素类型，避免 union 退化。
 */
type ListEnvelope<T> = { items: T[]; total: number };

export async function listTags(token: string, q?: string): Promise<ListEnvelope<TagView>> {
  const query = q && q.trim() ? `?q=${encodeURIComponent(q.trim())}` : '';
  return request<ListEnvelope<TagView>>(`/api/tags${query}`, { token });
}

export async function getTag(token: string, tagId: number): Promise<TagView> {
  return request<TagView>(`/api/tags/${tagId}`, { token });
}

export async function createTag(
  token: string,
  payload: { name: string; color?: string; description?: string },
): Promise<TagView> {
  return request<TagView>('/api/tags', {
    method: 'POST',
    token,
    body: JSON.stringify(payload),
  });
}

export async function updateTag(
  token: string,
  tagId: number,
  payload: { name?: string; color?: string; description?: string; status?: string },
): Promise<TagView> {
  return request<TagView>(`/api/tags/${tagId}`, {
    method: 'PUT',
    token,
    body: JSON.stringify(payload),
  });
}

export async function archiveTag(token: string, tagId: number): Promise<TagView> {
  return request<TagView>(`/api/tags/${tagId}`, {
    method: 'DELETE',
    token,
  });
}

export async function listDocumentTags(token: string, documentId: number): Promise<DocumentTagView[]> {
  const result = await request<ListEnvelope<DocumentTagView>>(`/api/documents/${documentId}/tags`, { token });
  return result.items;
}

export async function addDocumentTag(
  token: string,
  documentId: number,
  tagId: number,
): Promise<TagLinkView> {
  return request<TagLinkView>('/api/documents/${documentId}/tags', {
    method: 'POST',
    token,
    body: JSON.stringify({ tag_id: tagId }),
  });
}

export async function removeDocumentTag(token: string, documentId: number, tagId: number): Promise<void> {
  await request<{ deleted: boolean }>(`/api/documents/${documentId}/tags/${tagId}`, {
    method: 'DELETE',
    token,
  });
}

export async function listDocumentsByTag(token: string, tagId: number): Promise<KnowledgeDocument[]> {
  const result = await request<ListEnvelope<KnowledgeDocument>>(`/api/tags/${tagId}/documents`, { token });
  return result.items;
}

File: backend/api/tags.py
"""FastAPI router for REQ-012 tags (API-014/027/031/032, Phase2A minimal).

GET/POST /api/tags；GET/PUT/DELETE /api/tags/{id}；
GET/POST /api/documents/{id}/tags；DELETE /api/documents/{id}/tags/{tag_id}；
GET /api/tags/{id}/documents。全部带空间隔离与文档权限过滤（service 层）。
单 router、全路径：因混 /api/tags 与 /api/documents/{id}/tags 两类根，不便用统一 prefix。
"""

from __future__ import annotations

from backend.repository import repository
from backend.service.auth_context import TokenContext, get_current_user
from backend.service.document import DocumentNotFoundError
from backend.service.tag import (
    DocumentTagView,
    Tag,
    TagCreateRequest,
    TagUpdateRequest,
    TagView,
    add_document_tag,
    archive_tag,
    create_tag,
    get_tag_detail,
    list_document_tags,
    list_documents_by_tag,
    list_tags,
    remove_document_tag,
    update_tag,
)

try:
    from fastapi import APIRouter, Depends, HTTPException
    from pydantic import BaseModel
except ImportError:  # pragma: no cover - allows service tests before dependencies are installed
    APIRouter = None
    BaseModel = object
    HTTPException = Exception


if APIRouter is not None:
    router = APIRouter(tags=["tags"])

    class TagCreateBody(BaseModel):
        name: str
        color: str | None = None
        description: str | None = None

    class TagUpdateBody(BaseModel):
        name: str | None = None
        color: str | None = None
        description: str | None = None
        status: str | None = None

    class DocumentTagCreateBody(BaseModel):
        tag_id: int

    @router.get("/api/tags")
    def list_tags_endpoint(
        q: str | None = None,
        status: str | None = "active",
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        views = list_tags(repository, ctx.user_id, ctx.current_space_id, q, status)
        items = [_tag_view(view) for view in views]
        return {"code": 0, "msg": "ok", "data": {"items": items, "total": len(items)}}

    @router.post("/api/tags")
    def create_tag_endpoint(
        request: TagCreateBody,
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        tag = create_tag(
            repository,
            ctx.user_id,
            ctx.current_space_id,
            TagCreateRequest(name=request.name, color=request.color, description=request.description),
        )
        return {"code": 0, "msg": "ok", "data": _tag_detail(tag)}

    @router.get("/api/tags/{tag_id}")
    def get_tag_endpoint(tag_id: int, ctx: TokenContext = Depends(get_current_user)) -> dict[str, object]:
        view = get_tag_detail(repository, ctx.user_id, ctx.current_space_id, tag_id)
        return {"code": 0, "msg": "ok", "data": _tag_view(view)}

    @router.put("/api/tags/{tag_id}")
    def update_tag_endpoint(
        tag_id: int,
        request: TagUpdateBody,
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        tag = update_tag(
            repository,
            ctx.user_id,
            ctx.current_space_id,
            tag_id,
            TagUpdateRequest(
                name=request.name,
                color=request.color,
                description=request.description,
                status=request.status,
            ),
        )
        return {"code": 0, "msg": "ok", "data": _tag_detail(tag)}

    @router.delete("/api/tags/{tag_id}")
    def archive_tag_endpoint(tag_id: int, ctx: TokenContext = Depends(get_current_user)) -> dict[str, object]:
        tag = archive_tag(repository, ctx.user_id, ctx.current_space_id, tag_id)
        return {"code": 0, "msg": "ok", "data": _tag_detail(tag)}

    @router.get("/api/documents/{document_id}/tags")
    def list_document_tags_endpoint(
        document_id: int,
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        try:
            views = list_document_tags(repository, ctx.user_id, ctx.current_space_id, document_id)
        except DocumentNotFoundError as exc:
            raise HTTPException(status_code=404, detail={"code": 4004, "msg": "document not found"}) from exc
        items = [_document_tag_view(view) for view in views]
        return {"code": 0, "msg": "ok", "data": {"items": items, "total": len(items)}}

    @router.post("/api/documents/{document_id}/tags")
    def add_document_tag_endpoint(
        document_id: int,
        request: DocumentTagCreateBody,
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        try:
            link = add_document_tag(
                repository, ctx.user_id, ctx.current_space_id, document_id, request.tag_id
            )
        except DocumentNotFoundError as exc:
            raise HTTPException(status_code=404, detail={"code": 4004, "msg": "document not found"}) from exc
        return {
            "code": 0,
            "msg": "ok",
            "data": {
                "tag_id": link.tag_id,
                "document_id": link.document_id,
                "link_source": link.link_source,
            },
        }

    @router.delete("/api/documents/{document_id}/tags/{tag_id}")
    def remove_document_tag_endpoint(
        document_id: int,
        tag_id: int,
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        try:
            remove_document_tag(repository, ctx.user_id, ctx.current_space_id, document_id, tag_id)
        except DocumentNotFoundError as exc:
            raise HTTPException(status_code=404, detail={"code": 4004, "msg": "document not found"}) from exc
        return {"code": 0, "msg": "ok", "data": {"deleted": True}}

    @router.get("/api/tags/{tag_id}/documents")
    def list_documents_by_tag_endpoint(
        tag_id: int,
        status: str | None = "active",
        ctx: TokenContext = Depends(get_current_user),
    ) -> dict[str, object]:
        documents = list_documents_by_tag(
            repository, ctx.user_id, ctx.current_space_id, tag_id, status
        )
        items = [_document_view(document) for document in documents]
        return {"code": 0, "msg": "ok", "data": {"items": items, "total": len(items)}}

    def _tag_view(view: TagView) -> dict[str, object]:
        return {
            "id": view.id,
            "name": view.name,
            "color": view.color,
            "description": view.description,
            "document_count": view.document_count,
            "status": view.status,
        }

    def _tag_detail(tag: Tag) -> dict[str, object]:
        return {
            "id": tag.id,
            "name": tag.name,
            "color": tag.color,
            "description": tag.description,
            "status": tag.status,
        }

    def _document_tag_view(view: DocumentTagView) -> dict[str, object]:
        return {
            "tag_id": view.tag_id,
            "name": view.name,
            "color": view.color,
            "link_source": view.link_source,
        }

    def _document_view(document) -> dict[str, object]:
        return {
            "id": document.id,
            "title": document.title,
            "permission": document.permission.value,
            "owner_id": document.owner_id,
            "updated_at": document.updated_at,
        }

else:
    router = None
